### ISSUE TITLE
No compilation error raised when Map in Struct

### ISSUE THREAD

**Issue by henry-hz** [vyperlang/vyper#1966]
### Version Information

* vyper Version (output of `vyper --version`): 0.1.0b17+commit.eaa6754
* OS: Debian GNU/Linux 10 (buster)
* Python Version - 3.6.2
* Environment (output of `pip freeze`):

asttokens==2.0.3
pkg-resources==0.0.0
pycryptodome==3.9.7
six==1.14.0
vyper==0.1.0b17

### What's your issue about?

Can't populate a single struct without adding an inside mapping when using a struct declaration variable inside a function

```
struct Teddy:
    id: uint256
    year: uint256
    component: map(uint256, uint256)

teddies: map(uint256, Teddy)

@public
def __init__():
    self.teddies[0].id = 0
    self.teddies[0].year = 2019

@public
def create_teddy(teddy_name: string[20]):
  teddy: Teddy = Teddy({
        id: 3,
        year: 1922
      })
```

output:
```
vyper.exceptions.InvalidType: line 4:15 No mappings allowed for in-memory types, only fixed-size arrays
     3     year: uint256
---> 4     component: map(uint256, uint256)
----------------------^
```

### How can it be fixed?

tip: strange, cleaning the function 'create_teddy' you just compile it... the context is affecting how the compiler 'understand' the struct declaration. I also tried to:

```
 self.teddies[0] = Teddy({
        id: 3,
        year: 1922
      })
```

but got a 
```
vyper.exceptions.TypeMismatch: line 15:2 Keys don't match for structs, missing component
```

**Comment by fubuloubu** [comment]
I'm not sure I understand the issue.

It is true that we currently disallow a struct from containing an inner mapping, so the `vyper.exceptions.InvalidType` error makes sense. It is also true that you cannot initialize a struct without all of it's members present, so the `vyper.exceptions.TypeMismatch` also makes sense.

Is your bug about the relative ordering of these two issues, and how they change when you add a new function? Ordering of error reporting is not guaranteed if you change the contents of the source file, certain checks happen in different stages so one may occur earlier than another check.

**Comment by henry-hz** [comment]
Thanks for the answer,
interesting, seems that it's not clear in the docs, and people think that inner mapping is okay:
https://ethereum.stackexchange.com/questions/74467/vyper-mapping-inside-a-struct

In fact I do compile without all members, the error only appear when more conditions are in the same time:
1. struct with inner map
2. populate the struct using MyStruct({bla: bla})

but if I don't use inner map, or if I use the myStruct[key] = value way, the compilers gives me an okay to move ahead, as in the example below. If it's not a bug, it is at least ambiguous.

```
struct Policy:                                                       
  name:             string[20]                                                            
  designer:         address               
  premium:          uint256
  agreed:           uint256                    
  payback_peridod:  uint256               
  accepted_delay:   uint256             
         

@public
def create_policy(name: string[20], premium: uint256, agreed: uint256) -> bytes32:
  hash: bytes32 = keccak256(name)
  self.policies[hash] = Policy({
        name: name,
        designer: msg.sender,
        premium: premium,
        agreed: agreed
      })
  return hash


                                                  
```

**Comment by fubuloubu** [comment]
So, a few things:
1. Maps of maps are fine (e.g. the Computable Labs example)
2. Your example with `Policy` is fine, `string[20]` can be a member of a struct as it has a defined size
3. When you set something indirectly via `myStruct[key] = value` it will indeed not cause a problem because you are not trying to initialize a struct containing unsized data
4. It is indeed a bug that a `map` can be a member of a struct, we didn't have a check for that but we are working on a refactor where we can add a check
5. I agree that there are ambiguities to this that should be addressed in the documentation so people are aware of what works and what doesn't

**Comment by henry-hz** [comment]
Thanks for the feedback @fubuloubu . In fact, I will miss the feature that "a map can't be a member of a struct"... that is found in solidity. I you can, please, could you explain me what was the design decision to take it out, and how would you rewrite the, for example,  the '    mapping(address => uint256) balances; ' nested in the struct below [in solidity] ?
```
  struct Circle {
    address owner;
    string name;
    address[] members;
    uint256 currentIndex;
    uint256 depositAmount;
    address tokenAddress;
    uint256 timestamp;
    mapping(address => uint256) balances;
  }
```
 in [this](https://github.com/celo-org/savings-circle-demo/blob/master/contracts/SavingsCircle.sol) contract ?

**Comment by fubuloubu** [comment]
Allowing a member of undefined size (a mapping) in a struct makes it unsized, meaning it cannot be used for operations that require a known size such as using it as a function argument, event arg, etc. To allow structs to have undefined size members, we would have to add a lot of complex checking around _how_ the struct would be used, throwing if it were used in any one of these scenarios (hopefully referencing the original definition so it's clear to the user what needs to be fixed). It has an easy work around, so it's not been a priority to add this functionality, although you're more than welcome to submit a PR. In your example:
```python
struct Circle:
    owner: address
    string name: string[N]  # You need to specify N 
    # we don't have a way of specifying dynamic arrays yet
    members: address[N]  # You need to specify N
    currentIndex: uint256
    depositAmount: uint256
    tokenAddress: address
    timestamp: uint256

# Key of this mapping is Circle.owner
circleBalances: map(address, uint256)
```

The trick is to use an important, unique member of the `Circle` struct to serve as the lookup key in `circleBalances`. It requires a bit of thought, but you can probably shave a lot of the extra members in `Circle` with a bit of thought on how they're used.

**Comment by henry-hz** [comment]
Thanks @fubuloubu for dedicating time to help me, I will stick with this approach you suggested.

**Comment by iamdefinitelyahuman** [comment]
Closed in #2027
